### PR TITLE
CORTX-28168: Implement health event generator "publish" option

### PR DESCRIPTION
**Pull request by indrajitzagade**
Signed-off-by: Indrajit Zagade <indrajit.zagade@seagate.com>

# Problem Statement
Implement health event generator "publish" option

# Design
https://seagate-systems.atlassian.net/wiki/spaces/PUB/pages/891781277/Action+Status+Update+LLD

# Coding
-  [x] Coding conventions are followed and code is consistent

# Testing 
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Review Checklist 
- [x] PR is self reviewed
- [x] JIRA number/GitHub Issue added to PR
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [x] Check if the description is clear and explained
- [x] Is there a change in filename/package/module or signature? [Y/N]: NA
- [x] If yes for above point, is a notification sent to all other cortx components? [Y/N]: NA
- [x] Side effects on other features (deployment/upgrade)? [Y/N]: NA
- [x] Dependencies on other component(s)? [Y/N]: NA
-     If yes for above point, post link to the corresponding PR.

# Review Checklist 
- [x] Is perfline test run and the report with and without the changes updated in the PR? [Y/N]: NA

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

-----
[View rendered ha/util/health_generator/README.md](https://github.com/indrajitzagade/cortx-ha/blob/event_generator/ha/util/health_generator/README.md)

# Unit Testing Details:

Node health status before publishing the health event:
cortx/ha/v1/cortx/ha/system/cluster/30a58c57a8ca4c2abd55ba857bf5f026/site/1/rack/1/node/658d34feae294a6197ec9f309dc846ef/health:{"events": [{"event_timestamp": "1646131702", "created_timestamp": "1646131702", "status": "unknown", "specific_info": null}], "action": {"modified_timestamp": "1646131702", "status": "pending"}, "attributes": {}}

Published the health event:
[root@cortx-ha-headless-svc /]# ./mock_health_event_publisher.py publish -f config.json
Publishing health event {'event': {'header': {'version': '1.0', 'timestamp': '1646382881', 'event_id': '1646382881e0da79eafb3046cba5f4a049028a5f13'}, 'payload': {'source': 'hare', 'cluster_id': '30a58c57a8ca4c2abd55ba857bf5f026', 'site_id': '1', 'rack_id': '1', 'storageset_id': '1', 'node_id': '658d34feae294a6197ec9f309dc846ef', 'resource_type': 'node', 'resource_id': '658d34feae294a6197ec9f309dc846ef', 'resource_status': 'online', 'specific_info': {}}}}

Node health status after publishing the health event:
cortx/ha/v1/cortx/ha/system/cluster/30a58c57a8ca4c2abd55ba857bf5f026/site/1/rack/1/node/658d34feae294a6197ec9f309dc846ef/health:{"events": [{"event_timestamp": "1646382881", "created_timestamp": "1646382882", "status": "online", "specific_info": {}}, {"event_timestamp": "1646131702", "created_timestamp": "1646131702", "status": "unknown", "specific_info": null}], "action": {"modified_timestamp": "1646382882", "status": "pending"}, "attributes": {}}

### CODE REVIEW

File: ha/util/health_generator/README.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    get-disks           Displys the Disk Ids assosciated with the Node
+    get-cvgs            Displys the CVG Ids assosciated with the Node
+    help                see `help -h`
+
+optional arguments:
+  -h, --help            show this help message and exit
+  -gdt, --get-data-nodes
+                        Get the list of data node ids
+  -gs, --get-server-nodes
+                        Get the list of server node ids
```
Comment (ArchanaLimaye) [inline ha/util/health_generator/README.md:23]: - can we have a few samples of the full commands to be executed on command line  here ?

- Also restrictions/limitations should be added somewhere right ? ex This script needs to be executed from HA container only etc

File: ha/util/health_generator/mock_health_event_publisher.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+                cluster_id = Conf.get(const.HA_GLOBAL_INDEX, f'COMMON_CONFIG{const._DELIM}cluster_id')
+                site_id = Conf.get(const.HA_GLOBAL_INDEX, f'COMMON_CONFIG{const._DELIM}site_id')
+                rack_id = Conf.get(const.HA_GLOBAL_INDEX, f'COMMON_CONFIG{const._DELIM}rack_id')
+                storageset_id = '1' # TODO: Read from config when available.
+                for _, value in events_dict[_events_key].items():
+                    resource_type = value[_resource_type_key]
+                    if resource_type not in ['node', 'cvg', 'disk']:
+                        raise Exception(f'Invalid resource_type: {resource_type}')
+                    resource_status = value[_resource_status_key]
+                    if resource_status not in ['recovering', 'online', 'failed', 'unknown', 'degraded', 'repairing', 'repaired', 'rebalancing', 'offline']:
```
Comment (ArchanaLimaye) [inline ha/util/health_generator/mock_health_event_publisher.py:133]: Should we use constants defined in HA here and for resource types  ?

File: ha/util/health_generator/mock_health_event_publisher.py [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+    publishes the message on the message bus.
+    Args:
+    conf_store: ConftStoreSearch object
+    node_id: machine_id value
+    config_file: config file path
+    """
+    try:
+        with open(args.file, 'r') as fi:
+            events_dict = json.load(fi)
+            if _events_key in events_dict.keys():
```
Comment (ArchanaLimaye) [inline ha/util/health_generator/mock_health_event_publisher.py:119]: If multiple events need to be sent, publish function /mock_health_event_publisher.py should be called multiple times right ?

Comment (indrajitzagade) [reply ha/util/health_generator/mock_health_event_publisher.py:119]: Yes, it is called in a for loop.

File: ha/util/health_generator/README.md (hunk 1 above)
Comment (indrajitzagade) [reply ha/util/health_generator/README.md:23]: Added.

File: ha/util/health_generator/mock_health_event_publisher.py (hunk 2 above)
Comment (indrajitzagade) [reply ha/util/health_generator/mock_health_event_publisher.py]: Done, reading supported resource types from HA config file now.

File: ha/util/health_generator/mock_health_event_publisher.py [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+                MessageBus.init()
+                message_type = Conf.get(const.HA_GLOBAL_INDEX, f'FAULT_TOLERANCE{const._DELIM}message_type')
+                message_producer = MessageBus.get_producer("health_event_generator", message_type)
+                cluster_id = Conf.get(const.HA_GLOBAL_INDEX, f'COMMON_CONFIG{const._DELIM}cluster_id')
+                site_id = Conf.get(const.HA_GLOBAL_INDEX, f'COMMON_CONFIG{const._DELIM}site_id')
+                rack_id = Conf.get(const.HA_GLOBAL_INDEX, f'COMMON_CONFIG{const._DELIM}rack_id')
+                storageset_id = '1' # TODO: Read from config when available.
+                for _, value in events_dict[_events_key].items():
+                    resource_type = value[_resource_type_key]
+                    if resource_type not in ['node', 'cvg', 'disk']:
```
[the remark this answers is not in the document]
Comment (indrajitzagade) [reply ha/util/health_generator/mock_health_event_publisher.py]: Reading and using from HA config file.

File: ha/util/health_generator/mock_health_event_publisher.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (indrajitzagade) [reply ha/util/health_generator/mock_health_event_publisher.py]: Done

File: ha/util/health_generator/mock_health_event_publisher.py [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+import errno
+import json
+import time
+
+from cortx.utils.conf_store import Conf
+from ha.util.conf_store import ConftStoreSearch
+from ha.core.config.config_manager import ConfigManager
+from ha import const
+from ha.util.message_bus import MessageBus
+from ha.fault_tolerance.event import Event
```
Comment (mariyappanp) [inline ha/util/health_generator/mock_health_event_publisher.py:34]: Still using Event class of HA? This is going to be removed with PR https://github.com/Seagate/cortx-ha/pull/662.
Can we use health event from utils? @indrajitzagade

**Review by ArchanaLimaye — APPROVED** [review]

File: ha/util/health_generator/mock_health_event_publisher.py (hunk 5 above)
Comment (indrajitzagade) [reply ha/util/health_generator/mock_health_event_publisher.py:34]: The utils PR has just been merged, let me update this.

Comment (indrajitzagade) [reply ha/util/health_generator/mock_health_event_publisher.py]: Done @mariyappanp

**Review by mariyappanp — APPROVED** [review]
LGTM.

**Review by Madhura-08 — APPROVED** [review]

**Review by akash2144 — APPROVED** [review]

**Review by shaileshsvc — APPROVED** [review]
Approved

**Review by mukhtar-inamdar — APPROVED** [review]

File: ha/util/health_generator/config.json [hunk 6]
```diff
@@ -0,0 +1,13 @@
+{
+    "events": {
```
[the remark this answers is not in the document]
Comment (indrajitzagade) [reply ha/util/health_generator/config.json:2]: I think it is ok to add here @mukhtar-inamdar

[the remark this answers is not in the document]
Comment (mukhtar-inamdar) [reply ha/util/health_generator/config.json:2]: ok

**Pull request merged**